enzyme/Enzyme/MLIR/Implementations/LinalgAutoDiffOpInterfaceImpl.cpp

```diff
@@ -0,0 +1,237 @@
+
+using namespace mlir;
+using namespace mlir::enzyme;
+
+namespace {
+
+Value invertMemref(Value inp, OpBuilder &builder, Location loc) {
+  MemRefType iType = dyn_cast<MemRefType>(inp.getType());
+  SmallVector<Value> dims;
+  SmallVector<Value> dimSubOnes;
+  SmallVector<Value> strides;
+  Value negOne = builder.create<arith::ConstantIndexOp>(loc, -1);
+  for (int i = 0; i < iType.getShape().size(); i++) {
+    Value dim = builder.create<memref::DimOp>(loc, inp, i);
+    dims.push_back(dim);
+    auto dimSubOne = builder.create<arith::AddIOp>(loc, dim, negOne);
+    dimSubOnes.push_back(dimSubOne);
+    strides.push_back(negOne);
+  }
+  Value view = builder.create<memref::SubViewOp>(
+      loc, inp, ValueRange(dimSubOnes), ValueRange(dims), ValueRange(strides));
+  return view;
+}
+// TODO: Try to find a way to register autodiff interface for all LinalgOps
+struct GenericOpInterfaceReverse
+    : public ReverseAutoDiffOpInterface::ExternalModel<
+          GenericOpInterfaceReverse, linalg::GenericOp> {
+  void createReverseModeAdjoint(Operation *op, OpBuilder &builder,
+                                MGradientUtilsReverse *gutils,
+                                SmallVector<Value> caches) const {
+    // TODO: currently in progress, this doesn't work
+    auto linalgOp = cast<linalg::LinalgOp>(op);
+    assert(linalgOp.hasBufferSemantics() &&
+           "Linalg op with tensor semantics not yet supported");
+    linalg::LinalgOp newOp =
+        cast<linalg::LinalgOp>(gutils->getNewFromOriginal(linalgOp));
+
+    auto cacheBuilder = OpBuilder(newOp);
+
+    // get iteration domain
+    AffineMap aMap = newOp.getShapesToLoopsMap();
+    SmallVector<Value> dims;
+    for (OpOperand *input : newOp.getInputOperands()) {
+      auto shape = cast<MemRefType>(input->get().getType()).getShape();
+      for (int i = 0; i < shape.size(); i++) {
+        auto dimI =
+            cacheBuilder.create<arith::ConstantIndexOp>(op->getLoc(), i);
+        // input->get().getParentBlock()->dump();
+        auto dim = cacheBuilder.create<memref::DimOp>(op->getLoc(),
+                                                      input->get(), dimI);
+        dims.push_back(dim);
+      }
+    }
+
+    SmallVector<Value> iterationDomains;
+    std::vector<int64_t> shapes;
+    for (unsigned int i = 0; i < aMap.getNumResults(); i++) {
+      AffineMap subMap = aMap.getSubMap({i});
+      Value domain = cacheBuilder.create<AffineApplyOp>(op->getLoc(), subMap,
+                                                        ValueRange(dims));
+      iterationDomains.push_back(domain);
+      shapes.push_back(-1);
+    }
+
+    SmallVector<Value> inputs, outputs;
+    SmallVector<AffineMap> indexingMaps;
+    SmallVector<StringRef> iteratorTypes{linalgOp.getNumLoops(),
+                                         getParallelIteratorTypeName()};
+
+    for (OpOperand *output : linalgOp.getOutputOperands()) {
+      if (gutils->hasInvertPointer(output->get())) {
```

Comment:
Nit: use `continue` here to simplify code https://llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code.